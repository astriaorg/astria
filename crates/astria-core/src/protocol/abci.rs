use std::{
    borrow::Cow,
    num::NonZeroU32,
};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[allow(clippy::module_name_repetitions)]
pub struct AbciErrorCode(u32);

#[rustfmt::skip]
impl AbciErrorCode {
    pub const UNSPECIFIED: Self = Self(0);
    pub const UNKNOWN_PATH: Self = Self(1);
    pub const INVALID_PARAMETER: Self = Self(2);
    pub const INTERNAL_ERROR: Self = Self(3);
    pub const INVALID_NONCE: Self = Self(4);
    pub const TRANSACTION_TOO_LARGE: Self = Self(5);
    pub const INSUFFICIENT_FUNDS: Self = Self(6);
    pub const INVALID_CHAIN_ID: Self = Self(7);
}

impl AbciErrorCode {
    #[must_use]
    pub fn info(self) -> Cow<'static, str> {
        match self.0 {
            0 => "unspecified".into(),
            1 => "provided path is unknown".into(),
            2 => "one or more path parameters were invalid".into(),
            3 => "an internal server error occurred".into(),
            4 => "the provided nonce was invalid".into(),
            5 => "the provided transaction was too large".into(),
            6 => "insufficient funds".into(),
            7 => "the provided chain id was invalid".into(),
            other => format!("unknown non-zero abci error code: {other}").into(),
        }
    }
}

impl std::fmt::Display for AbciErrorCode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.info())
    }
}

impl From<AbciErrorCode> for tendermint::abci::Code {
    fn from(value: AbciErrorCode) -> Self {
        value.0.into()
    }
}

impl From<NonZeroU32> for AbciErrorCode {
    fn from(value: NonZeroU32) -> Self {
        match value.get() {
            1 => Self::UNKNOWN_PATH,
            2 => Self::INVALID_PARAMETER,
            3 => Self::INTERNAL_ERROR,
            4 => Self::INVALID_NONCE,
            5 => Self::TRANSACTION_TOO_LARGE,
            6 => Self::INSUFFICIENT_FUNDS,
            7 => Self::INVALID_CHAIN_ID,
            other => Self(other),
        }
    }
}
