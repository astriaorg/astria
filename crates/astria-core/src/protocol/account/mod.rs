pub mod v1alpha1;

use crate::generated::protocol::account::v1alpha1 as raw;
