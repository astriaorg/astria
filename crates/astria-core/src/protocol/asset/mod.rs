pub mod v1alpha1;

use crate::generated::protocol::asset::v1alpha1 as raw;
