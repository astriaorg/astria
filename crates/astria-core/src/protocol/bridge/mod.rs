pub mod v1;

use crate::generated::protocol::bridge::v1 as raw;
