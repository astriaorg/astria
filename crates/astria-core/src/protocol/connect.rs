pub mod v1 {
    use indexmap::IndexMap;
    use tendermint::{
        abci::types::ExtendedCommitInfo,
        block::Round,
    };

    use crate::{
        connect::types::v2::{
            CurrencyPair,
            CurrencyPairError,
            CurrencyPairId,
        },
        generated::{
            astria::protocol::connect::v1 as raw,
            astria_vendored,
        },
    };

    impl From<astria_vendored::tendermint::abci::ExtendedCommitInfo>
        for tendermint_proto::abci::ExtendedCommitInfo
    {
        fn from(value: astria_vendored::tendermint::abci::ExtendedCommitInfo) -> Self {
            tendermint_proto::abci::ExtendedCommitInfo {
                round: value.round,
                votes: value.votes.into_iter().map(Into::into).collect(),
            }
        }
    }

    impl From<astria_vendored::tendermint::abci::ExtendedVoteInfo>
        for tendermint_proto::abci::ExtendedVoteInfo
    {
        fn from(value: astria_vendored::tendermint::abci::ExtendedVoteInfo) -> Self {
            tendermint_proto::abci::ExtendedVoteInfo {
                validator: value.validator.map(Into::into),
                vote_extension: value.vote_extension,
                extension_signature: value.extension_signature,
                block_id_flag: value.block_id_flag,
            }
        }
    }

    impl From<astria_vendored::tendermint::abci::Validator> for tendermint_proto::abci::Validator {
        fn from(value: astria_vendored::tendermint::abci::Validator) -> Self {
            tendermint_proto::abci::Validator {
                address: value.address,
                power: value.power,
            }
        }
    }

    impl From<tendermint_proto::abci::ExtendedCommitInfo>
        for astria_vendored::tendermint::abci::ExtendedCommitInfo
    {
        fn from(value: tendermint_proto::abci::ExtendedCommitInfo) -> Self {
            astria_vendored::tendermint::abci::ExtendedCommitInfo {
                round: value.round,
                votes: value.votes.into_iter().map(Into::into).collect(),
            }
        }
    }

    impl From<tendermint_proto::abci::ExtendedVoteInfo>
        for astria_vendored::tendermint::abci::ExtendedVoteInfo
    {
        fn from(value: tendermint_proto::abci::ExtendedVoteInfo) -> Self {
            astria_vendored::tendermint::abci::ExtendedVoteInfo {
                validator: value.validator.map(Into::into),
                vote_extension: value.vote_extension,
                extension_signature: value.extension_signature,
                block_id_flag: value.block_id_flag,
            }
        }
    }

    impl From<tendermint_proto::abci::Validator> for astria_vendored::tendermint::abci::Validator {
        fn from(value: tendermint_proto::abci::Validator) -> Self {
            astria_vendored::tendermint::abci::Validator {
                address: value.address,
                power: value.power,
            }
        }
    }

    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
    pub struct CurrencyPairInfo {
        pub currency_pair: CurrencyPair,
        pub decimals: u64,
    }

    impl std::fmt::Display for CurrencyPairInfo {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            write!(
                f,
                "CurrencyPairInfo {{ currency_pair: {}, decimals: {} }}",
                self.currency_pair, self.decimals
            )
        }
    }

    #[derive(Debug, Clone)]
    pub struct ExtendedCommitInfoWithCurrencyPairMapping {
        pub extended_commit_info: ExtendedCommitInfo,
        pub id_to_currency_pair: IndexMap<CurrencyPairId, CurrencyPairInfo>,
    }

    impl ExtendedCommitInfoWithCurrencyPairMapping {
        #[must_use]
        pub fn new(
            extended_commit_info: ExtendedCommitInfo,
            id_to_currency_pair: IndexMap<CurrencyPairId, CurrencyPairInfo>,
        ) -> Self {
            Self {
                extended_commit_info,
                id_to_currency_pair,
            }
        }

        #[must_use]
        pub fn empty(round: Round) -> Self {
            Self {
                extended_commit_info: ExtendedCommitInfo {
                    round,
                    votes: Vec::new(),
                },
                id_to_currency_pair: IndexMap::new(),
            }
        }

        /// Converts from a protobuf `ExtendedCommitInfoWithCurrencyPairMapping` to the native type.
        ///
        /// # Errors
        ///
        /// - if the `extended_commit_info` field is not set
        /// - if the `extended_commit_info` field is invalid
        /// - if a `currency_pair` field is not set within a `id_to_currency_pair` item
        /// - if a currency pair is invalid
        pub fn try_from_raw(
            raw: raw::ExtendedCommitInfoWithCurrencyPairMapping,
        ) -> Result<Self, ExtendedCommitInfoWithCurrencyPairMappingError> {
            let Some(extended_commit_info) = raw.extended_commit_info else {
                return Err(
                    ExtendedCommitInfoWithCurrencyPairMappingError::field_not_set(
                        "extended_commit_info",
                    ),
                );
            };
            let extended_commit_info = ExtendedCommitInfo::try_from(
                tendermint_proto::abci::ExtendedCommitInfo::from(extended_commit_info),
            )
            .map_err(ExtendedCommitInfoWithCurrencyPairMappingError::extended_commit_info)?;
            let id_to_currency_pair = raw
                .id_to_currency_pair
                .into_iter()
                .map(|id_with_currency_pair| {
                    let currency_pair_id = CurrencyPairId::new(id_with_currency_pair.id);
                    let Some(currency_pair) = id_with_currency_pair.currency_pair else {
                        return Err(
                            ExtendedCommitInfoWithCurrencyPairMappingError::field_not_set(
                                "currency_pair",
                            ),
                        );
                    };
                    let currency_pair = CurrencyPair::try_from_raw(currency_pair)
                        .map_err(ExtendedCommitInfoWithCurrencyPairMappingError::currency_pair)?;
                    Ok((
                        currency_pair_id,
                        CurrencyPairInfo {
                            currency_pair,
                            decimals: id_with_currency_pair.decimals,
                        },
                    ))
                })
                .collect::<Result<
                    IndexMap<CurrencyPairId, CurrencyPairInfo>,
                    ExtendedCommitInfoWithCurrencyPairMappingError,
                >>()?;
            Ok(Self {
                extended_commit_info,
                id_to_currency_pair,
            })
        }

        #[must_use]
        pub fn into_raw(self) -> raw::ExtendedCommitInfoWithCurrencyPairMapping {
            let extended_commit_info: tendermint_proto::abci::ExtendedCommitInfo =
                self.extended_commit_info.into();
            let id_to_currency_pair = self
                .id_to_currency_pair
                .into_iter()
                .map(
                    |(currency_pair_id, currency_pair_info)| raw::IdWithCurrencyPair {
                        id: currency_pair_id.get(),
                        currency_pair: Some(currency_pair_info.currency_pair.into_raw()),
                        decimals: currency_pair_info.decimals,
                    },
                )
                .collect();
            raw::ExtendedCommitInfoWithCurrencyPairMapping {
                extended_commit_info: Some(extended_commit_info.into()),
                id_to_currency_pair,
            }
        }
    }

    #[derive(Debug, thiserror::Error)]
    #[error(transparent)]
    pub struct ExtendedCommitInfoWithCurrencyPairMappingError(
        ExtendedCommitInfoWithCurrencyPairMappingErrorKind,
    );

    impl ExtendedCommitInfoWithCurrencyPairMappingError {
        fn field_not_set(field: &'static str) -> Self {
            Self(
                ExtendedCommitInfoWithCurrencyPairMappingErrorKind::FieldNotSet {
                    field,
                },
            )
        }

        fn currency_pair(error: CurrencyPairError) -> Self {
            Self(ExtendedCommitInfoWithCurrencyPairMappingErrorKind::CurrencyPair(error))
        }

        fn extended_commit_info(error: tendermint::Error) -> Self {
            Self(ExtendedCommitInfoWithCurrencyPairMappingErrorKind::ExtendedCommitInfo(error))
        }
    }

    #[derive(Debug, thiserror::Error)]
    enum ExtendedCommitInfoWithCurrencyPairMappingErrorKind {
        #[error("field not set: {field}")]
        FieldNotSet { field: &'static str },
        #[error("invalid currency pair")]
        CurrencyPair(#[from] CurrencyPairError),
        #[error("invalid extended commit info")]
        ExtendedCommitInfo(#[from] tendermint::Error),
    }
}
