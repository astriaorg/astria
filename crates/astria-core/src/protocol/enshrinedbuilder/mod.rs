mod v1alpha1;
