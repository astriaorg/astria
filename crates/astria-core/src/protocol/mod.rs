use indexmap::IndexMap;
use transaction::v1alpha1::SignedTransaction;

use crate::primitive::v1::RollupId;

pub mod abci;
pub mod account;
pub mod asset;
pub mod transaction;

#[cfg(any(feature = "test-utils", test))]
pub mod test_utils;

/// Extracts all data within [`SequenceAction`]s in the given [`SignedTransaction`]s, wraps them as
/// [`RollupData::SequencedData`] and groups them by [`RollupId`].
///
/// TODO: This can all be done in-place once <https://github.com/rust-lang/rust/issues/80552> is stabilized.
pub fn group_sequence_actions_in_signed_transaction_transactions_by_rollup_id(
    signed_transactions: &[SignedTransaction],
) -> IndexMap<RollupId, Vec<Vec<u8>>> {
    use prost::Message as _;

    use crate::sequencerblock::v1alpha1::block::RollupData;

    let mut map = IndexMap::new();
    for action in signed_transactions
        .iter()
        .flat_map(SignedTransaction::actions)
    {
        if let Some(action) = action.as_sequence() {
            let txs_for_rollup: &mut Vec<Vec<u8>> = map.entry(action.rollup_id).or_insert(vec![]);
            let rollup_data = RollupData::SequencedData(action.data.clone());
            txs_for_rollup.push(rollup_data.into_raw().encode_to_vec());
        }
    }
    map.sort_unstable_keys();
    map
}
