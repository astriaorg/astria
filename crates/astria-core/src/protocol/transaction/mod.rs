pub mod v1alpha1;

use crate::generated::protocol::transaction::v1alpha1 as raw;
