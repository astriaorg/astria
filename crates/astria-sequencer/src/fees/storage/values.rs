use astria_core::protocol::{
    fees::v1::FeeComponents as DomainFeeComponents,
    transaction::v1::action::{
        BridgeLock,
        BridgeSudoChange,
        BridgeUnlock,
        FeeAssetChange,
        FeeChange,
        IbcRelayerChange,
        IbcSudoChange,
        Ics20Withdrawal,
        InitBridgeAccount,
        RollupDataSubmission,
        SudoAddressChange,
        Transfer,
        ValidatorUpdate,
    },
};
use astria_eyre::eyre::bail;
use borsh::{
    BorshDeserialize,
    BorshSerialize,
};
use penumbra_ibc::IbcRelay;

#[derive(Debug, BorshSerialize, BorshDeserialize)]
pub(crate) struct Value(ValueImpl);

#[derive(Debug, BorshSerialize, BorshDeserialize)]
#[expect(
    clippy::enum_variant_names,
    reason = "want to make it clear that these are fees and not actions"
)]
enum ValueImpl {
    TransferFees(FeeComponents),
    SequenceFees(FeeComponents),
    Ics20WithdrawalFees(FeeComponents),
    InitBridgeAccountFees(FeeComponents),
    BridgeLockFees(FeeComponents),
    BridgeUnlockFees(FeeComponents),
    BridgeSudoChangeFees(FeeComponents),
    IbcRelayFees(FeeComponents),
    ValidatorUpdateFees(FeeComponents),
    FeeAssetChangeFees(FeeComponents),
    FeeChangeFees(FeeComponents),
    IbcRelayerChangeFees(FeeComponents),
    IbcSudoChangeFees(FeeComponents),
    SudoAddressChangeFees(FeeComponents),
}

macro_rules! impl_from_for_fee_storage {
    ( $( $domain_ty:ty => $value_impl:ident),* $(,)? ) => {
        $(
            impl<'a> From<$domain_ty> for crate::storage::StoredValue<'a> {
                fn from(fees: $domain_ty) -> Self {
                    crate::storage::StoredValue::Fees(Value(ValueImpl::$value_impl(fees.into())))
                }
            }
            impl<'a> TryFrom<crate::storage::StoredValue<'a>> for $domain_ty {
                type Error = astria_eyre::eyre::Error;

                fn try_from(value: crate::storage::StoredValue<'a>) -> Result<Self, Self::Error> {
                    let crate::storage::StoredValue::Fees(Value(ValueImpl::$value_impl(fees))) = value else {
                        let value_impl_ty = concat!("ValueImpl::", stringify!($value_impl));
                        bail!(
                            "fees stored value type mismatch: expected {value_impl_ty}, found {value:?}"
                        );
                    };
                    Ok(fees.into())
                }
            }
        )*
    }
}

#[derive(Debug, BorshSerialize, BorshDeserialize)]
pub(in crate::fees) struct FeeComponents {
    pub base: u128,
    pub multiplier: u128,
}

impl<T> From<DomainFeeComponents<T>> for FeeComponents {
    fn from(fees: DomainFeeComponents<T>) -> Self {
        Self {
            base: fees.base(),
            multiplier: fees.multiplier(),
        }
    }
}

impl<T: ?Sized> From<FeeComponents> for DomainFeeComponents<T> {
    fn from(fees: FeeComponents) -> Self {
        Self::new(fees.base, fees.multiplier)
    }
}

impl_from_for_fee_storage!(
    DomainFeeComponents<Transfer> => TransferFees,
    DomainFeeComponents<RollupDataSubmission> => SequenceFees,
    DomainFeeComponents<Ics20Withdrawal> => Ics20WithdrawalFees,
    DomainFeeComponents<InitBridgeAccount> => InitBridgeAccountFees,
    DomainFeeComponents<BridgeLock> => BridgeLockFees,
    DomainFeeComponents<BridgeUnlock> => BridgeUnlockFees,
    DomainFeeComponents<BridgeSudoChange> => BridgeSudoChangeFees,
    DomainFeeComponents<IbcRelay> => IbcRelayFees,
    DomainFeeComponents<ValidatorUpdate> => ValidatorUpdateFees,
    DomainFeeComponents<FeeAssetChange> => FeeAssetChangeFees,
    DomainFeeComponents<FeeChange> => FeeChangeFees,
    DomainFeeComponents<IbcRelayerChange> => IbcRelayerChangeFees,
    DomainFeeComponents<IbcSudoChange> => IbcSudoChangeFees,
    DomainFeeComponents<SudoAddressChange> => SudoAddressChangeFees,
);
