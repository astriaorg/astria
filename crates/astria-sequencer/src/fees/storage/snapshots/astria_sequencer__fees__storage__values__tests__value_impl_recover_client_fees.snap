---
source: crates/astria-sequencer/src/fees/storage/values.rs
expression: value_impl_borsh_as_hex!(RecoverIbcClientFees)
---
0f0100000000000000000000000000000002000000000000000000000000000000
